insights/parsers/mysqladmin.py

```diff
@@ -15,6 +15,38 @@
 from insights.specs import Specs
 
 
+@parser(Specs.mysqladmin_status)
+class MysqladminStatus(LegacyItemAccess, CommandParser):
+    """
+    Module for parsing the output of the ``mysqladmin status`` command.
+    Typical output looks like::
+        Uptime: 1103965 Threads: 1820 Questions: 44778091 Slow queries: 0 Opens: 1919 Flush tables: 1 Open tables: 592 Queries per second avg: 40.561
+    Example:
+        >>> "Uptime" in mysqlstat
+        True
+        >>> mysqlstat['Threads']
+        '1820'
+        >>> mysqlstat['Flush_tables'] == '1'
+        True
+    """
+    def parse_content(self, content):
+        if not content:
+            raise ParseException("Input content is empty.")
+        self.data = {}
```

Comment:
@asyedham why not call this attribute `self.status` instead of the less descriptive `self.data`?

Reply:
@bafhr When i use self.status instead of self.data getting the following errors.

`Failed` example:
    "Uptime" in mysqlstat
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib64/python3.7/doctest.py", line 1329, in __run
        compileflags, 1), test.globs)
      File "<doctest insights.parsers.mysqladmin.MysqladminStatus[0]>", line 1, in <module>
        "Uptime" in mysqlstat
      File "/home/asma/new-insights-core/insights-core/insights/core/__init__.py", line 430, in __contains__
        return item in self.data
    AttributeError: 'MysqladminStatus' object has no attribute 'data'
**********************************************************************
File "/home/asma/new-insights-core/insights-core/insights/parsers/mysqladmin.py", line 33, in insights.parsers.mysqladmin.MysqladminStatus
Failed example:
    mysqlstat['Threads']
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib64/python3.7/doctest.py", line 1329, in __run
        compileflags, 1), test.globs)
      File "<doctest insights.parsers.mysqladmin.MysqladminStatus[1]>", line 1, in <module>
        mysqlstat['Threads']
      File "/home/asma/new-insights-core/insights-core/insights/core/__init__.py", line 427, in __getitem__
        return self.data[item]
    AttributeError: 'MysqladminStatus' object has no attribute 'data'
**********************************************************************
File "/home/asma/new-insights-core/insights-core/insights/parsers/mysqladmin.py", line 35, in insights.parsers.mysqladmin.MysqladminStatus
Failed example:
    mysqlstat['Flush_tables'] == '1'
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib64/python3.7/doctest.py", line 1329, in __run
        compileflags, 1), test.globs)
      File "<doctest insights.parsers.mysqladmin.MysqladminStatus[2]>", line 1, in <module>
        mysqlstat['Flush_tables'] == '1'
      File "/home/asma/new-insights-core/insights-core/insights/core/__init__.py", line 427, in __getitem__
        return self.data[item]
    AttributeError: 'MysqladminStatus' object has no attribute 'data'
**********************************************************************
1 items had failures:
   3 of   3 in insights.parsers.mysqladmin.MysqladminStatus
***Test Failed*** 3 failures.
FAILED
=========================================================================================== short test summary info ===========================================================================================
FAIL insights/parsers/tests/test_mysqladmin.py::test_mysqladmin_status
FAIL insights/parsers/tests/test_mysqladmin.py::test_doc

================================================================================================== FAILURES ===================================================================================================
___________________________________________________________________________________________ test_mysqladmin_status ____________________________________________________________________________________________

    def test_mysqladmin_status():
        mysqlstat = MysqladminStatus(context_wrap(OUTPUT_MYSQLADMIN_STATUS))
>       assert "Uptime" in mysqlstat

insights/parsers/tests/test_mysqladmin.py:22: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <insights.parsers.mysqladmin.MysqladminStatus object at 0x7f821d55e390>, item = 'Uptime'

    def __contains__(self, item):
>       return item in self.data
E       AttributeError: 'MysqladminStatus' object has no attribute 'data'

insights/core/__init__.py:430: AttributeError
__________________________________________________________________________________________________ test_doc ___________________________________________________________________________________________________

    def test_doc():
        env = {
                'MysqladminStatus': MysqladminStatus,
                'mysqlstat': MysqladminStatus(context_wrap(OUTPUT_MYSQLADMIN_STATUS, path='/bin/mysqladmin status')),
                'MysqladminVars': MysqladminVars,
                'output': MysqladminVars(context_wrap(INPUT_NORMAL, '/bin/mysqladmin variables')),
        }
        failed, total = doctest.testmod(mysqladmin, globs=env)
>       assert failed == 0
E       AssertionError

insights/parsers/tests/test_mysqladmin.py:175: AssertionError

Reply:
@asyedham the reason that happens is because you are using the `LegacyItemAccess` mixin class which expects there to be a `self.data` dictionary.  You would need to make additional changes to make this change work.

Reply:
@asyedham It makes more sense to use `self.status` attrib instead of `self.data` as suggested by @bfahr. Make sure you don't inherit from `LegacyItemAccess`.

Furthermore, as the output is defined and the elements are documented in the docs[1] why not make use to Python's regex to map key-value pairs? Below is the sample implementations:

```python
import re
from itertools import zip_longest
from insights import CommandParser, parser
from insights.parsers import ParseException, SkipException
from insights.specs import Specs


@parser(Specs.mysqladmin_status)
class MysqladminStatus(CommandParser):
    """
    Module for parsing the output of the ``mysqladmin status`` command.

    Typical output looks like:

        Uptime: 1103965 Threads: 1820 Questions: 44778091 Slow queries: 0 Opens: 1919 Flush tables: 1 Open tables: 592 Queries per second avg: 40.561

    Examples:
        >>> "Uptime" in mysqlstat
        True
        >>> mysqlstat['Threads']
        '1820'
        >>> mysqlstat['Flush_tables'] == '1'
        True
    """
    pattern = re.compile("(Uptime):.(\d+).(Threads):.(\d+).(Questions):.(\d+).(Slow queries):.(\d+).(Opens):.(\d+).(Flush tables):.(\d+).(Open tables):.(\d+).(Queries per second avg):.(\d+.\d*)")

    def grouper(self, iterable, n, fillvalue=None):
        "Collect data into fixed-length chunks or blocks"
        args = [iter(iterable)] * n
        return zip_longest(*args, fillvalue=fillvalue)

    def parse_content(self, content):
        if not content:
            raise SkipException("Content is empty.")

        self.status = {}
        if self.pattern.search(content[0]):
            groups = self.pattern.search(content[0]).groups()
            for group in self.grouper(groups, 2):
                self.status[group[0]] = group[1]
        else:
            raise ParseException("Unable to parse the output.")

```


[1] https://dev.mysql.com/doc/refman/5.7/en/mysqladmin.html

Reply:
@**psachin** Done